### PR TITLE
Vm get exercise

### PR DESCRIPTION
**Pull request by JungAn2**
Finished starting the vM.
The ip and port will not be there since there is some issue with qemu-agent not being able to install to some vulnerable image.
There will be another issue for novnc.

### CODE REVIEW

File: config/app.ts [hunk 1]
```diff
@@ -2,9 +2,18 @@
...
-	port: process.env.PORT ? process.env.PORT : 3000,
-	ldap: process.env.LDAP ? process.env.LDAP : "IDK"
-}
\ No newline at end of file
+  port: process.env.PORT ? process.env.PORT : 3000,
+  ldap: process.env.LDAP ? process.env.LDAP : "IDK",
+  ticket_id: process.env.TICKET_ID ? process.env.TICKET_ID : "wrong",
+  ticket_uuid: process.env.TICKET_UUID ? process.env.TICKET_UUID : "hahaha",
+  url: process.env.URL ? process.env.URL : "move along",
+  nodename: process.env.NODENAME ? process.env.NODENAME : "don't mind me",
```
Comment (Desss64) [inline config/app.ts:13]: I'm not sure what "IDK", "move along", "don't mind me" etc do in this block but I'm sure it's ok lol.

**Review by Desss64 — APPROVED** [review]
Everything looks good to me

File: @types/express/index.d.ts [hunk 2]
```diff
@@ -1,9 +1,12 @@
+import { VM } from ".prisma/client";
+
 export {};
 
 declare global {
   namespace Express {
     interface Request {
       user?: User;
+      VM: VM;
```
Comment (wrathofrathma) [inline @types/express/index.d.ts:9]: 👀 😠

File: app/services/VirtualMachineService.ts [hunk 3]
```diff
@@ -0,0 +1,584 @@
...
+export const VirtualMachineService = {
+  /**
+   * Creates the VM in proxmox
+   *
+   * @param {string} user current user
+   * @param {string} exerciseId vmid of exercise
+   * @param {string} nodeName node name to manual assign load
+   * @return error from function
+   */
+  async createVM(
```
Comment (wrathofrathma) [inline app/services/VirtualMachineService.ts:22]: Make sure to validate the node sent by the user.

I want to maintain an array of "whitelisted" nodes on the backend. Those are the ONLY nodes the client is aware of. 

Which means on the frontend, when you select a node to launch the VM on, it should source those nodes from this list of whitelisted nodes.

File: app/services/VirtualMachineService.ts [hunk 4]
```diff
@@ -0,0 +1,584 @@
...
+          exerciseId: exerciseId,
+          ip: "",
+          port: "",
+          timeLimit: metadata["timelimit"],
+          timeEnd: endTime,
+          status: "running",
+        },
+      });
+    } else {
+      return await prisma.vM.update({
```
Comment (wrathofrathma) [inline app/services/VirtualMachineService.ts:146]: Test if you can make the prisma.vm lowercase or uppercase. If you can make it a consistent casing, do that.

**Review by wrathofrathma — CHANGES_REQUESTED** [review]
fix me 🏗️ 🔨

File: app/services/VirtualMachineService.ts (hunk 3 above)
Comment (wrathofrathma) [reply app/services/VirtualMachineService.ts:22]: I also want aliases.

File: app/services/VirtualMachineService.ts (hunk 4 above)
Comment (JungAn2) [reply app/services/VirtualMachineService.ts:150]: Doing it all uppercase works. Will change all the vm prisma to uppercases

**Review by wrathofrathma — APPROVED** [review]

**Pull request merged**